sdk/python/feast/infra/offline_stores/bigquery.py

```diff
@@ -206,15 +207,35 @@ def _infer_event_timestamp_from_dataframe(entity_df: pandas.DataFrame) -> str:
 
 
 class BigQueryRetrievalJob(RetrievalJob):
-    def __init__(self, query, client):
+    def __init__(self, query, client, config):
         self.query = query
         self.client = client
+        self.config = config
 
     def to_df(self):
         # TODO: Ideally only start this job when the user runs "get_historical_features", not when they run to_df()
         df = self.client.query(self.query).to_dataframe(create_bqstorage_client=True)
         return df
 
+    def to_bigquery(self, dry_run=False) -> str:
+        path = f"{self.client.project}.{self.config.offline_store.dataset}.tmp-{str(uuid.uuid4())}"
+        job_config = bigquery.QueryJobConfig(destination=path, dry_run=dry_run)
+        bq_job = self.client.query(self.query, job_config=job_config)
+
+        if dry_run is True:
+            print("This query will process {} bytes.".format(bq_job.total_bytes_processed))
+            return
+
+        while True:
+            query_job = self.client.get_job(bq_job.job_id)
```

Comment:
I don't understand what this line of code does. What is the difference between the first job and the second job? Isn't it the same job?

Reply:
Yeah this is to bypass some strange behaviour -- without this, the job state just by using `result.state()` or `.done()` doesn't refresh properly. Encountered this both on Matt's and my end.

Reply:
> the job state just by using result.state() or .done() doesn't refresh properly

Indeed ! The state does not refresh automatically. So the best practice is to fetch the job through the API over and over ...
If you know a better way to achieve that, please let us know 🙏

Reply:
Also, if we are going to write retry logic, should we just bite the bullet and use https://tenacity.readthedocs.io/en/latest/? I would rather make our code easier to read at the expense of dependencies, and I can see us needing retries in other places as well.

Reply:
Tenacity looks nice. Thanks for sharing

For this use case, I would rather use BQ functions (like `done()`) as it's available. No need to add dependencies for now if that works

Reply:
> `done()`is supposed to work if you use reload=True

As per the docs you linked, reload=True is default so this would've been the case when we tried.

Reply:
To use the `done()` capabilities, we need to permissions too `bigquery.readsessions.create`

![image](https://user-images.githubusercontent.com/18557047/122085328-9ca6c980-cdd0-11eb-8132-fab19dab5a45.png)